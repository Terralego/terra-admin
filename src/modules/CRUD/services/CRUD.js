import Api from '@terralego/core/modules/Api';
import { sanitizeCustomEndpoint } from './utils';

export const ACTION_CREATE = 'create';

const snakeToCamel = str => str.replace(
  /([-_][a-z])/g,
  group => group.toUpperCase().substring(1),
);

const camelCaseKeys = obj => Object.keys(obj).reduce((props, key) => ({
  ...props,
  [snakeToCamel(key)]: obj[key],
}), {});


export const fetchSettings = endpoint =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}/`);


const flattenMenu = menu => (
  menu.reduce((list, { crud_views: views }) => (
    [...list, ...views]
  ), [])
);

export const getView = ({ menu = [] } = {}, value, layerField = 'layer.name') => {
  const view = flattenMenu(menu);
  if (!view.length || value === undefined) {
    return false;
  }
  const viewProps = view.find(item => {
    const propToFind = layerField.split('.').reduce((obj, prop) => obj?.[prop], item);
    return propToFind === value;
  }) || {};

  return viewProps.layer
    ? camelCaseKeys(viewProps)
    : false;
};

export const getFirstCrudViewName = (settings = {}) => {
  const { menu: [{ crud_views: [{ layer: { name } = {} } = {}] = [] } = {}] = [] } = settings;
  return name;
};

export const getSources = ({ menu = [] }) =>
  flattenMenu(menu).reduce((sourceList, { layer }) => (
    (sourceList.some(({ id }) => id === `${layer.id}`))
      ? sourceList
      : [
        ...sourceList,
        { id: `${layer.id}`, type: 'vector', url: layer.tilejson },
      ]
  ), []);

export const getLayers = ({ menu = [] }) =>
  flattenMenu(menu).reduce((
    layersList,
    {
      layer: { id },
      map_layers: mapLayers,
    },
  ) => ([
    ...layersList,
    ...mapLayers.map(({ id_layer_vt: name, style, ...props }) => ({
      id: `CRUD-${name}-${id}`,
      'source-layer': name,
      source: `${id}`,
      ...style,
      ...props,
    })),
  ]), []);

export default { fetchSettings, getView, getSources, getLayers, getFirstCrudViewName };
