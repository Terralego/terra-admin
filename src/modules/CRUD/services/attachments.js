import Api from '@terralego/core/modules/Api';
import { sanitizeCustomEndpoint } from './utils';

const attachmentRequest = (endpoint, body, method = 'PUT') =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}/`, { method, body });

export const saveAttachment = attachmentRequest;

export const deleteAttachment = endpoint =>
  attachmentRequest(endpoint, {}, 'DELETE');

export const saveAttachmentCategories = body =>
  Api.request('crud/attachment-categories/', { method: 'POST', body });
