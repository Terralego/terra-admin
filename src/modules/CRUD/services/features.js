import Api from '@terralego/core/modules/Api';
import { sanitizeCustomEndpoint } from './utils';

export const fetchFeaturesList = (endpoint, querystring = {}) =>
  Api.request(
    `${sanitizeCustomEndpoint(endpoint)}`,
    { querystring },
  );

export const fetchFeature = (endpoint, featureId) =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}/${featureId}/`);

const createFeature = (endpoint, body) =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}/`, { method: 'POST', body });

const updateFeature = (endpoint, featureId, body, method = 'PUT') =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}/${featureId}/`, { method, body });

export const deleteFeature = (endpoint, featureId) =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}/${featureId}/`, { method: 'DELETE' });

export const saveFeature = (endpoint, featureId, body, method) => (
  featureId
    ? updateFeature(endpoint, featureId, body, method)
    : createFeature(endpoint, body)
);

export const fetchCustomEndpoint = (endpoint, querystring) =>
  Api.request(`${sanitizeCustomEndpoint(endpoint)}`, querystring);

export const updateFeatureIdentifier = (feature, newFeature) => ({
  ...feature,
  ...(newFeature.identifier
    ? { [newFeature.identifier]: newFeature }
    : {}
  ),
});

const getBoundingBox = (list, item) => {
  const [
    [minLng, minLat],
    [maxLng, maxLat],
  ] = list;

  const [lng, lat] = item;

  return [
    [Math.min(minLng, lng), Math.min(minLat, lat)],
    [Math.max(maxLng, lng), Math.max(maxLat, lat)],
  ];
};

export const getBounds = (coordinates, limits = [[Infinity, Infinity], [-Infinity, -Infinity]]) => {
  if (!Array.isArray(coordinates[0])) {
    return getBoundingBox(limits, coordinates);
  }
  return coordinates.reduce((list, coordinate) => {
    if (!Array.isArray(coordinate[0])) {
      return getBoundingBox(list, coordinate);
    }
    return getBounds(coordinate, list);
  }, limits);
};


export default {
  fetchFeaturesList,
  fetchFeature,
  deleteFeature,
  saveFeature,
  fetchCustomEndpoint,
  getBounds,
  updateFeatureIdentifier,
};
