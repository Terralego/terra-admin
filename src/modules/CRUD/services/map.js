import config from './mock-config.json';

// Mock api the config return
export const fetchMapConfig = () => ({ ...config });

export default { fetchMapConfig };
