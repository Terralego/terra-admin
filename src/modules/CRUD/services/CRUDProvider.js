import React from 'react';
import connect from 'react-ctx-connect';
import { fetchSettings } from './CRUD';
import {
  fetchFeaturesList,
  fetchFeature as fetchFeatureAction,
  saveFeature as saveFeatureAction,
  deleteFeature as deleteFeatureAction,
  updateFeatureIdentifier,
  fetchCustomEndpoint,
} from './features';

import { saveAttachmentCategories } from './attachments';

export const CRUDContext = React.createContext({});
export const connectCRUDProvider = connect(CRUDContext);

const { Provider } = CRUDContext;

const PAGE_SIZE = 999;

export class CRUDProvider extends React.Component {
  state = {
    settings: undefined,
    featuresList: {},
    feature: {},
    mapConfig: undefined,
    errors: {
      attachmentCategories: undefined,
      settings: undefined,
      featuresList: [],
      feature: [],
    },
  };

  dataTableRef = React.createRef();

  detailsRef = React.createRef();

  componentWillUnmount () {
    this.isUnmount = true;
  }

  setMap = map => !this.isUnmount && this.setState({ map });

  getSettings = async endpoint => {
    const result = {};
    try {
      const settings = await fetchSettings(endpoint);
      result.settings = settings;
    } catch (e) {
      result.error = e;
    }

    const { settings = {}, error } = result;

    this.setState(({ errors }) => ({
      settings,
      errors: {
        ...errors,
        settings: error,
      },
    }));

    return settings;
  }

  getFormattedError = ({
    store,
    error,
    ids,
  }) => {
    const { errors: { [store]: errorStore } } = this.state;

    const propsFromIds = Object.keys(ids);

    // Remove error if it was before in the store
    if (!(error instanceof Error)) {
      return errorStore.filter(errorItem => (
        propsFromIds.every(prop => errorItem[prop] !== ids[prop])
      ));
    }

    // Modify error if it was before in the store
    const isErrorItemAlreadyExisting = errorStore.some(errorItem => (
      propsFromIds.every(prop => errorItem[prop] === ids[prop])
    ));

    if (isErrorItemAlreadyExisting) {
      return errorStore.map(errorItem => (
        propsFromIds.every(prop => errorItem[prop] === ids[prop])
          ? { error, ...ids }
          : errorItem
      ));
    }

    // Else add error in the store
    return [...errorStore, { error, ...ids }];
  }

  getFeaturesList = async (layerId, querystring) => {
    const result = {};
    try {
      const featuresList = await fetchFeaturesList(layerId, querystring);
      result.featuresList = featuresList;
    } catch (e) {
      result.error = e;
    }
    const { featuresList = [], error = {} } = result;

    this.setState(({ errors }) => ({
      featuresList,
      errors: {
        ...errors,
        featuresList: this.getFormattedError({ error, ids: { layerId }, store: 'featuresList' }),
      },
    }));
    return featuresList.length && featuresList;
  };

  fetchFeature = async (layerId, featureId) => {
    const result = {};
    try {
      const feature = await fetchFeatureAction(layerId, featureId);
      result.feature = feature;
    } catch (e) {
      result.error = e;
    }

    const { feature: newFeature = {}, error = {} } = result;

    this.setState(({ errors, feature }) => ({
      feature: updateFeatureIdentifier(feature, newFeature),
      errors: {
        ...errors,
        feature: this.getFormattedError({ error, ids: { layerId, featureId }, store: 'feature' }),
      },
    }));

    return Object.keys(newFeature).length > 0 && newFeature;
  }

  saveFeature = async (layerId, featureId, data, method = 'PUT') => {
    const result = {};
    try {
      const feature = await saveFeatureAction(layerId, featureId, data, method);
      result.feature = feature;
    } catch (e) {
      result.error = e;
    }

    const { feature: newFeature = {}, error = {} } = result;

    this.setState(({ errors, feature }) => ({
      feature: updateFeatureIdentifier(feature, newFeature),
      errors: {
        ...errors,
        feature: this.getFormattedError({ error, ids: { layerId, featureId }, store: 'feature' }),
      },
    }));

    return result;
  }

  deleteFeature = async (layerId, featureId) => {
    const result = {};
    try {
      await deleteFeatureAction(layerId, featureId);
      result.feature = featureId;
    } catch (e) {
      // Error
    }
    const { feature } = result;

    this.setState(({ featuresList, feature: { [feature]: deletedFeature, ...rest } }) => ({
      feature: rest,
      featuresList: {
        ...featuresList,
        results: deletedFeature
          ? featuresList.results.filter(({ identifier }) => identifier !== featureId)
          : featuresList.results,
      },
    }));

    return feature;
  }

  getAttachmentCategories = async endpoint => {
    const { attachmentCategories } = this.state;
    if (attachmentCategories) {
      return attachmentCategories;
    }
    const result = {};
    try {
      const categories = await fetchCustomEndpoint(`${endpoint}`, {
        querystring: { page_size: PAGE_SIZE },
      });
      result.attachmentCategories = categories;
    } catch (e) {
      result.error = e;
    }

    const { attachmentCategories: nextAttachmentCategories, error } = result;

    this.setState(({ errors }) => ({
      attachmentCategories: nextAttachmentCategories,
      errors: {
        ...errors,
        attachmentCategories: error,
      },
    }));

    return nextAttachmentCategories;
  }

  createAttachmentCategories = async name => {
    const result = {};
    try {
      const category = await saveAttachmentCategories(name);
      result.category = category;
    } catch (e) {
      result.error = e;
    }

    const { category: newCategory = {}, error = {} } = result;

    this.setState(({ errors, attachmentCategories }) => ({
      attachmentCategories: {
        ...attachmentCategories,
        newCategory,
      },
      errors: {
        ...errors,
        attachmentCategories: {
          ...attachmentCategories.error,
          error,
        },
      },
    }));

    return Object.keys(newCategory).length > 0 && newCategory;
  }

  findOrCreateAttachmentCategory = async category => {
    const { name, id } = category;
    if (id !== null) {
      return category;
    }
    const {
      attachmentCategories = [],
      settings: {
        config: {
          attachment_categories: endpoint,
        },
      },
    } = this.state;
    let categories = attachmentCategories;
    if (!categories.length) {
      const { results } = await this.getAttachmentCategories(endpoint);
      categories = results;
    }
    const categoryInList = categories.find(cat => (
        cat?.name.toLowerCase() === name.trim().toLowerCase()
    ));
    if (categoryInList !== undefined) {
      return categoryInList;
    }
    const newCategory = await this.createAttachmentCategories(name);
    return newCategory;
  }

  render () {
    const { children } = this.props;

    const {
      dataTableRef,
      detailsRef,
      getMapConfig,
      getSettings,
      getFeaturesList,
      fetchFeature,
      saveFeature,
      deleteFeature,
      getAttachmentCategories,
      findOrCreateAttachmentCategory,
      setMap,
    } = this;
    const value = {
      ...this.state,
      dataTableRef,
      detailsRef,
      getMapConfig,
      getSettings,
      getFeaturesList,
      fetchFeature,
      saveFeature,
      deleteFeature,
      getAttachmentCategories,
      findOrCreateAttachmentCategory,
      setMap,
    };

    return (
      <Provider value={value}>
        {children}
      </Provider>
    );
  }
}

export default CRUDProvider;
