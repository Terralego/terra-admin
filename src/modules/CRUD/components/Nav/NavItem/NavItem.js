import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { Icon, Popover, PopoverInteractionKind, Position } from '@blueprintjs/core';
import { getBounds } from '../../../services/features';

import { generateURI } from '../../../config';

import NavIcon from '../NavIcon';

const NavAddFeature = ({ withPopover, children, t }) => {
  if (!withPopover) {
    return children;
  }

  return (
    <Popover
      content={t('CRUD.details.create')}
      interactionKind={PopoverInteractionKind.HOVER}
      position={Position.RIGHT}
      boundary="window"
    >
      {children}
    </Popover>
  );
};


const NavItem = ({
  dataTableRef,
  displayAddFeature,
  extent,
  layer,
  map,
  minified,
  name,
  pictogram,
  t,
}) => {
  const handleClick = ({ target }) => {
    const isActivePage = !!target.offsetParent.getAttribute('aria-current');
    if (!isActivePage || !map) {
      return;
    }
    const [w, s, e, n] = extent;
    const { current: dataTable } = dataTableRef;

    setTimeout(() => {
      map.resize();
      map.fitBounds(getBounds([[w, s], [e, n]]), {
        padding: {
          top: 20,
          right: 50,
          bottom: dataTable.offsetHeight + 20,
          left: 20,
        },
        duration: 0,
      });
    }, 600);
  };
  return (
    <div className="CRUD-nav__action">
      <NavLink
        className="CRUD-nav__link"
        to={generateURI('layer', { layer: layer.name })}
        onClick={handleClick}
      >
        <span className="CRUD-nav__item-content">
          <span className="bp3-button bp3-minimal">
            {!minified && <NavIcon src={pictogram} />}
            <span className="bp3-button-text">{name}</span>
          </span>
        </span>
      </NavLink>
      {displayAddFeature && (
        <NavAddFeature withPopover={!minified} t={t}>
          <NavLink to={generateURI('layer', { layer: layer.name, id: 'create' })}>
            <span className="CRUD-nav__item-content">
              <span className="bp3-button bp3-minimal">
                <Icon icon="add" color="#666" />
              </span>
            </span>
          </NavLink>
        </NavAddFeature>
      )}
    </div>
  );
};

export default NavItem;

NavItem.propTypes = {
  dataTableRef: PropTypes.shape({}),
  map: PropTypes.shape({}),
  name: PropTypes.string.isRequired,
  pictogram: PropTypes.string,
  layer: PropTypes.shape({
    name: PropTypes.string,
  }),
  minified: PropTypes.bool,
  displayAddFeature: PropTypes.bool,
  t: PropTypes.func,
};

NavItem.defaultProps = {
  dataTableRef: null,
  pictogram: undefined,
  layer: {
    name: undefined,
  },
  map: undefined,
  minified: false,
  displayAddFeature: false,
  t: text => text,
};
