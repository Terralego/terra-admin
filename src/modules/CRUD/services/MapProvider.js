import React, { useState, useCallback, useRef, useEffect } from 'react';
import connect from 'react-ctx-connect';
import { useTranslation } from 'react-i18next';

import {
  DEFAULT_CONTROLS,
  CONTROL_CAPTURE,
  CONTROL_BACKGROUND_STYLES,
  CONTROL_SEARCH,
  CONTROLS_TOP_RIGHT,
} from '@terralego/core/modules/Map';

import { getBounds } from './features';

const BASE_SEARCH_API_PROVIDER = 'https://nominatim.openstreetmap.org/search?';

const CONTROL_LIST = [{
  control: CONTROL_BACKGROUND_STYLES,
  position: CONTROLS_TOP_RIGHT,
}, {
  control: CONTROL_CAPTURE,
  position: CONTROLS_TOP_RIGHT,
}];

const sortByOrder = ({ order: a = 0 }, { order: b = 0 }) => a - b;

export const MapContext = React.createContext({});
export const connectMapProvider = connect(MapContext);

const { Provider } = MapContext;

export const MapProvider = ({ children }) => {
  const detailsRef = useRef(null);
  const dataTableRef = useRef(null);

  const [map, setMap] = useState(null);

  const { i18n } = useTranslation();

  const setFitBounds = useCallback(({ coordinates, hasDetails }) => {
    if (!coordinates.length || !map) return;

    const { current: detail } = detailsRef;
    const { current: dataTable } = dataTableRef;

    const padding = {
      top: 20,
      right: hasDetails ? (detail?.offsetWidth ?? 0 + 50) : 50,
      bottom: !hasDetails ? (dataTable?.offsetHeight ?? 0 + 20) : 20,
      left: 20,
    };
    map.resize();
    map.fitBounds(getBounds(coordinates), { padding, duration: 0 });
  }, [dataTableRef, detailsRef, map]);

  const onSearch = useCallback(async query => {
    const params = new URLSearchParams();
    params.append('q', query);
    params.append('format', 'geojson');
    params.append('accept-language', i18n.language);

    const headers = new Headers([['Content-Type', 'application/json']]);
    let results;
    try {
      results = await fetch(`${BASE_SEARCH_API_PROVIDER}${params}`, {
        headers,
      }).then(response => response.json());
    } catch (e) {
      return null;
    }
    // Filter to avoid duplicates location
    const filteredFeatures = results.features.reduce(
      (list, result) => {
        if (!list.some(item => item.properties.display_name === result.properties.display_name)) {
          list.push(result);
        }
        return list;
      }, [],
    );
    const data = [{
      total: filteredFeatures.length,
      results: filteredFeatures.map(
        ({ bbox, properties: { osm_id: id, display_name: label } }) => ({
          label,
          id,
          bounds: bbox,
        }),
      ),
    }];
    return data;
  }, [i18n.language]);

  const onSearchResultClick = useCallback(({ result: { bounds } }) => {
    const [xMin, yMin, xMax, yMax] = bounds;
    setFitBounds({
      coordinates: [[xMin, yMin], [xMax, yMax]],
      hasDetails: Boolean(detailsRef.current),
    });
  }, [setFitBounds]);

  const [controls, setControls] = useState([...DEFAULT_CONTROLS, ...CONTROL_LIST]);
  const [interactiveMapProps, setInteractiveMapProps] = useState(undefined);
  const [layers, setLayers] = useState([]);

  useEffect(() => {
    if (!map) {
      return;
    }
    setControls(prevControls => {
      if (prevControls.some(({ control }) => control === CONTROL_SEARCH)) {
        return prevControls;
      }
      return [{
        control: CONTROL_SEARCH,
        position: CONTROLS_TOP_RIGHT,
        maxResults: 10,
        onSearch,
        onSearchResultClick,
        order: -1,
      }, ...prevControls].sort(sortByOrder);
    });
  }, [map, onSearch, onSearchResultClick]);

  const addControl = useCallback(controlToAdd => {
    if (!controlToAdd) return;
    setControls(prevControls => {
      if (prevControls.some(({ control }) => control === controlToAdd.control)) {
        return prevControls;
      }
      return [controlToAdd, ...prevControls].sort(sortByOrder);
    });
  }, []);

  const removeControl = useCallback(controlToRemove => {
    if (!controlToRemove) return;
    setControls(prevControls => {
      if (prevControls.some(({ control }) => control === controlToRemove)) {
        return prevControls.filter(({ control }) => control !== controlToRemove);
      }
      return prevControls;
    });
  }, []);

  const featureToHighlight = useCallback(({ featureId, layer, hover }) => {
    const { id: layerId, source } = layers.find(({ 'source-layer': sourceLayer }) => sourceLayer === layer) || {};
    if (!map || !featureId || !layerId) {
      return;
    }
    const { addHighlight, removeHighlight } = interactiveMapProps;
    if (hover) {
      addHighlight && addHighlight({
        layerId,
        featureId,
        highlightColor: 'red',
        unique: true,
        source,
      });
    } else {
      removeHighlight && removeHighlight({
        layerId,
        featureId,
      });
    }
  }, [interactiveMapProps, layers, map]);


  const value = {
    addControl,
    controls,
    detailsRef,
    dataTableRef,
    featureToHighlight,
    layers,
    interactiveMapProps,
    map,
    setControls,
    setFitBounds,
    setMap,
    setLayers,
    setInteractiveMapProps,
    removeControl,
  };

  return (
    <Provider value={value}>
      {children}
    </Provider>
  );
};

export default MapProvider;
