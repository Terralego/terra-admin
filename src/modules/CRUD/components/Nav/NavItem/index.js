import { connectAuthProvider } from '@terralego/core/modules/Auth';
import compose from '../../../../../utils/compose';

import NavItem from './NavItem';

const authProviderGetter = ({
  authenticated,
  user,
}) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayAddFeature: permissions.includes('can_add_feature'),
  };
};

export default compose(
  connectAuthProvider(authProviderGetter),
)(NavItem);
