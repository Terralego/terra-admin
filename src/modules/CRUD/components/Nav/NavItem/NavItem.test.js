import React from 'react';
import renderer from 'react-test-renderer';
import { MapContext } from '../../../services/MapProvider';

import NavItem from './NavItem';

jest.mock('react-router-dom', () => ({
  NavLink: ({ children, ...props }) => <div {...props}>{children}</div>,
}));

jest.mock('@blueprintjs/core', () => ({
  Icon: ({ icon }) => <div>{icon}</div>,
  Popover: ({ children, ...props }) => <div {...props}>{children}</div>,
  PopoverInteractionKind: {
    HOVER: 'hover',
  },
  Position: {
    RIGHT: 'right',
  },
}));

jest.mock('../../../config', () => ({
  generateURI: jest.fn(),
}));

jest.mock('@terralego/core/modules/Map', () => ({}));

it('should snapshot correctly', () => {
  const tree = renderer.create((
    <MapContext.Provider value={{
      setFitBounds: jest.fn(),
    }}
    >
      <NavItem
        name="foo"
        displayAddFeature
        displayPictogram
        pictogram="path/to/picture"
      />
    </MapContext.Provider>
  )).toJSON();
  expect(tree).toMatchSnapshot();
});
