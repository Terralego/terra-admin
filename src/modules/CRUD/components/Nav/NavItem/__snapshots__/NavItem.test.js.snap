// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot correctly 1`] = `
<div
  className="CRUD-nav__action"
>
  <div
    className="CRUD-nav__link"
    onClick={[Function]}
  >
    <span
      className="CRUD-nav__item-content"
    >
      <span
        className="bp3-button bp3-minimal"
      >
        <span
          className=" bp3-icon"
        >
          <img
            alt=""
            className="CRUD-nav__icon"
            src="path/to/picture"
          />
        </span>
        <span
          className="bp3-button-text"
        >
          foo
        </span>
      </span>
    </span>
  </div>
  <div
    boundary="window"
    content="CRUD.details.create"
    interactionKind="hover"
    position="right"
  >
    <div>
      <span
        className="CRUD-nav__item-content"
      >
        <span
          className="bp3-button bp3-minimal"
        >
          <div>
            add
          </div>
        </span>
      </span>
    </div>
  </div>
</div>
`;
